### ISSUE TITLE
Rate limit for actions

### ISSUE THREAD

**Issue by jemand771** [OliveTin/OliveTin#76]
**Is your feature request related to a problem? Please describe.**
Currently, buttons that launch disruptive actions may be used to DoS the affected service. e.g. if I have a button for `systemctl restart docker` (for whatever reason), someone could repeatedly spam click it to permanently bring down all containers for as long as they keep clicking 

**Describe the solution you'd like**
I'd like a new config option that limits the number of executions of an action for a certain time interval. this should apply globally to all browsers/clients so you can't easily circumvent it using clientside mods or multiple browsers. nice to have: greyed out action button with a countdown overlay while the spam protection/cooldown is tripped. 

I haven't thought about the details too much yet, but there are probably a lot of ways to configure this. as a couple of examples:
* minimum delay between executions (primitive but very intuitive)
* rolling x executions per y time - there could also be more than one of these (e.g. max 2/hour but also 10/day)
* leaky bucket style rate+burst

**Describe alternatives you've considered**
* check the timeout (or whatever) inside a bash script, call that from OliveTin - the script will either abort early or go through and execute the actual command. this works well but is cumbersome to set up for many actions
* somehow prevent parallel executions of a single action and add a ` && sleep 300` to its command - _super_ hacky but would probably work too I guess..? (needs increased timeout in OliveTin)

**Additional context**
of course, I generally trust people I give access to OliveTin. nevertheless, adding some kind of spam prevention in case someone's account gets hacked or someone turns "evil" feels like a good idea. this can also be useful for preventing impatient elderly users from murdering my media server by pressing the rescan button a couple of times too many.

if this is out of scope because the project is aimed at a more "friendly" environment where no abuse prevention is needed, I can understand that too

**Comment by jamesread** [comment]
Hey @jemand771 , first of all, thanks for your time in writing this up. I too had thought about doing something like this before. 

I personally really like the "x per y" style configuration, it's pretty expressive. I think combining it with "maxConcurrentInstances" (which defaults to 1) would also be good. Something like;

````
actions:
    - name: run backup
      shell: /opt/backupScript.py
      maxConcurrentInstances; 1
      maxFrequency: 1 executions per 1 day
````

Do you think that could work? I'm not sure if this would create problems, or if people would need to specify multiple "frequencies" at a time (not sure how that would work out).

**Comment by jemand771** [comment]
I agree, the frequency-based approach is probably the easiest while still providing lots of control.
On setting multiple frequencies: I could see this being useful to prevent different "kinds of spam"
a) a short interval (e.g. once per minute) to counter impatience
b) a longer interval (e.g. twice per hour) to prevent (possibly accidental) abuse of an expensive operation
Some job might take a minute or two to show its effects (e.g. restarting a server), but also might need to be run multiple times within a short time frame _without_ allowing thousands of runs per day.
<details>
<summary>concrete examples from my setup, oddly specific</summary>
1. a discord bot
trusted moderators should get a way to restart a discord bot we're hosting. this can be a temporary fix for newly arisen issues that non-devs can apply quickly and easily. if a restart fails for some reason (e.g. external service temporarily unavailable), a second restart might be necessary. this means that a strict limit like once per day would be too agressive, but a higher one (e.g. 20) would allow 20 back-to-back restarts, causing significant downtime.

2. media server
media server users get a button to force a full library rescan. while you're ingesting media, this might need to be run multiple times (e.g. user forgot to add a file) -> pretty relaxed short-term limit (e.g. 1/minute). however, triggering a rescan _every minute of the day_ (which 1/m would allow) would cause too much strain on the underlying hardware in the long run.

again, I know these are oddly specific but I wanted to give you something practical to work with instead of just abstract "hey _maybe_ someone needs this" suggestions. the general theme is: job might need to run multiple times within a short time frame without allowing it to run too often overall
</details>

I also see that that would be more of an advanced feature and one frequency is probably enough for most use cases. nevertheless, syntax-wise I'm thinking of something like
```yaml
actions:
  - name: something expensive
    shell: /scripts/whatever.sh
    maxConcurrentInstances: 1
    rateLimitIntervals:
      - count: 1
        time: 1m
      - count: 2
        time: 1h
```
where `time` can be your generic duration-describing string. there are probably better names for this than `count` and `time`, but those were the first words that came to mind

**Comment by jamesread** [comment]
@jemand771 thanks for you very long patience :-) 

Implemented now, this will be in the 2024.04.18 release that is just building now.

Documented here; https://docs.olivetin.app/ratelimits.html
